### PR TITLE
Add regression test for #5621

### PR DESCRIPTION
**Pull request by thehesiod**
## What do these changes do?
Adds regression test for issue #5621

## Related issue number
#5621

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by thehesiod** [comment]
master version of https://github.com/aio-libs/aiohttp/pull/5633

**Comment by webknjaz** [comment]
@thehesiod could you drop those unrelated file changes with typings?

**Comment by webknjaz** [comment]
Also, the linter is unhappy :( Those annoying style issues should be fixed...

**Comment by thehesiod** [comment]
> @thehesiod could you drop those unrelated file changes with typings?

if I don't keep those there are PEP typing warnings in the file

**Comment by thehesiod** [comment]
sigh, not sure why mypy is complaining, pycharm is happy, seems like it doesn't recognize Union, I'll revert then and keep the warning

**Comment by thehesiod** [comment]
ah, figured it out

File: aiohttp/test_utils.py [hunk 1]
```diff
@@ -144,14 +144,14 @@ async def start_server(self, **kwargs: Any) -> None:
     async def _make_runner(self, **kwargs: Any) -> BaseRunner:
         pass
 
-    def make_url(self, path: str) -> URL:
+    def make_url(self, path: Union[str, URL]) -> URL:
```
Comment (webknjaz) [inline aiohttp/test_utils.py]: If these changes are necessary, please submit them in a separate PR with a separate changelog fragment.

File: tests/test_web_urldispatcher.py [hunk 2]
```diff
@@ -458,3 +459,47 @@ async def test_static_absolute_url(aiohttp_client: Any, tmp_path: Any) -> None:
...
+    raises=AssertionError,
+    reason="Regression in v3.7: https://github.com/aio-libs/aiohttp/issues/5621",
+)
+async def test_decoded_raw_match_regex(aiohttp_client) -> None:
+    app = web.Application()
+
+    async def handler(_):
+        return web.Response()
+
+    app.router.add_get("/467%2C802%2C24834%2C24952%2C25362%2C40574/hello", handler)
```
Comment (webknjaz) [inline tests/test_web_urldispatcher.py]: Why doesn't this one have a regex? If the reproducer doesn't depend on it, why don't we just remove it from the first test too? Also, the URL is quite long and since the test doesn't seem to be influenced by its length, let's just make it 2-character, one char would be ascii and the other one would be something urlencoded — this will reduce the cognitive load for the readers.

File: tests/test_web_urldispatcher.py [hunk 3]
```diff
@@ -458,3 +459,47 @@ async def test_static_absolute_url(aiohttp_client: Any, tmp_path: Any) -> None:
     client = await aiohttp_client(app)
     resp = await client.get("/static/" + str(file_path.resolve()))
     assert resp.status == 403
+
+
+@pytest.mark.xfail(
+    raises=AssertionError,
+    reason="Regression in v3.7: https://github.com/aio-libs/aiohttp/issues/5621",
+)
+async def test_decoded_match_regex(aiohttp_client) -> None:
```
Comment (webknjaz) [inline tests/test_web_urldispatcher.py]: Looking at this closer, I think it's actually one test with multiple params so let's just parametrize it:
```suggestion
@pytest.mark.parametrize(
    ("route_definition", "urlencoded_path", "expected_http_resp_status"),
    (
        ("/{user_ids:([0-9]+)(,([0-9]+))*}/hello", "/467%2C802%2C24834/hello", 200),
        ("/1%2C3/hello, "/1%2C3/hello", 404),
    ),
    ids=("urldecoded_route", "urlencoded_route"),
)
async def test_decoded_url_match(
    aiohttp_client,
    route_definition,
    urlencoded_path,
    expected_http_resp_status,
) -> None:
```

File: tests/test_web_urldispatcher.py [hunk 4]
```diff
@@ -458,3 +459,47 @@ async def test_static_absolute_url(aiohttp_client: Any, tmp_path: Any) -> None:
...
+    app = web.Application()
+
+    async def handler(_):
+        return web.Response()
+
+    app.router.add_get(
+        "/{user_ids:(([0-9]+)|(u_[0-9a-f]{16}))(,(([0-9]+)|(u_[0-9a-f]{16})))*}/hello",
+        handler,
+    )
```
Comment (webknjaz) [inline tests/test_web_urldispatcher.py]:
```suggestion
    app.router.add_get(route_definition, handler)
```

File: tests/test_web_urldispatcher.py [hunk 5]
```diff
@@ -458,3 +459,47 @@ async def test_static_absolute_url(aiohttp_client: Any, tmp_path: Any) -> None:
...
+    client = await aiohttp_client(app)
+
+    # '/467,802,24834,24952,25362,40574/hello'
+    r = await client.get(
+        yarl.URL("/467%2C802%2C24834%2C24952%2C25362%2C40574/hello", encoded=True)
+    )
```
Comment (webknjaz) [inline tests/test_web_urldispatcher.py]:
```suggestion
    r = await client.get(yarl.URL(urlencoded_path, encoded=True))
```

File: tests/test_web_urldispatcher.py [hunk 6]
```diff
@@ -458,3 +459,47 @@ async def test_static_absolute_url(aiohttp_client: Any, tmp_path: Any) -> None:
...
+    assert r.status == 200
```
Comment (webknjaz) [inline tests/test_web_urldispatcher.py]:
```suggestion
    assert r.status == expected_http_resp_status
```

File: tests/test_web_urldispatcher.py (hunk 2 above)
Comment (thehesiod) [reply tests/test_web_urldispatcher.py]: it doesn't need a regex, this is ensuring you can't have a match to an escaped path

Comment (webknjaz) [reply tests/test_web_urldispatcher.py]: I have a feeling that the first test doesn't need that either. Just register a decoded route, right?

File: tests/test_web_urldispatcher.py (hunk 3 above)
Comment (webknjaz) [reply tests/test_web_urldispatcher.py]: Or maybe even
```suggestion
@pytest.mark.parametrize(
    ("route_definition", "urlencoded_path", "expected_http_resp_status"),
    (
        ("/467,802,24834/hello", "/467%2C802%2C24834/hello", 200),
        ("/1%2C3/hello, "/1%2C3/hello", 404),
    ),
    ids=("urldecoded_route", "urlencoded_route"),
)
async def test_decoded_url_match(
    aiohttp_client,
    route_definition,
    urlencoded_path,
    expected_http_resp_status,
) -> None:
```

File: tests/test_web_urldispatcher.py (hunk 2 above)
Comment (thehesiod) [reply tests/test_web_urldispatcher.py]: correct, I've written it to match the actual bug in our production system.  I think it's worth keeping both ways to ensure both the regex works and the non regex, resulting in more code coverage

Comment (thehesiod) [reply tests/test_web_urldispatcher.py]: ok cool, I wasn't sure if you guys like separate tests or combo tests

Comment (webknjaz) [reply tests/test_web_urldispatcher.py]: Okay, then, please incorporate the earlier parametrization suggestions and keep both cases as params.

File: tests/test_web_urldispatcher.py (hunk 3 above)
Comment (webknjaz) [reply tests/test_web_urldispatcher.py]: Looking at this closer, I think it's actually one test with multiple params so let's just parametrize it:
```suggestion
@pytest.mark.parametrize(
    ("route_definition", "urlencoded_path", "expected_http_resp_status"),
    (
        ("/467,802,24834/hello", "/467%2C802%2C24834/hello", 200),
        ("/{user_ids:([0-9]+)(,([0-9]+))*}/hello", "/467%2C802%2C24834/hello", 200),
        ("/1%2C3/hello, "/1%2C3/hello", 404),
    ),
    ids=("urldecoded_route", "urldecoded_route_with_regex", "urlencoded_route"),
)
async def test_decoded_url_match(
    aiohttp_client,
    route_definition,
    urlencoded_path,
    expected_http_resp_status,
) -> None:
```

File: aiohttp/test_utils.py (hunk 1 above)
Comment (thehesiod) [reply aiohttp/test_utils.py]: removed

File: tests/test_web_urldispatcher.py (hunk 3 above)
Comment (webknjaz) [reply tests/test_web_urldispatcher.py]: @thehesiod with this suggestion, I believe you can just accept the patches via GH UI and then pull+tweak the other test removal locally.

File: CHANGES/5635.misc [hunk 7]
```diff
@@ -0,0 +1 @@
+Add regression test for match regex encoding.
```
Comment (webknjaz) [inline CHANGES/5635.misc]: I believe this would be more precise:
```suggestion
Added regression tests for dispatching urlencoded routes.
```

File: tests/test_web_urldispatcher.py [hunk 8]
```diff
@@ -458,3 +459,34 @@ async def test_static_absolute_url(aiohttp_client: Any, tmp_path: Any) -> None:
...
+
+
+@pytest.mark.xfail(
+    raises=AssertionError,
+    reason="Regression in v3.7: https://github.com/aio-libs/aiohttp/issues/5621",
+)
+@pytest.mark.parametrize(
+    ("route_definition", "urlencoded_path", "expected_http_resp_status"),
+    (
+        ("/{user_ids:([0-9]+)(,([0-9]+))*}/hello", "/467%2C802%2C24834/hello", 200),
```
[the remark this answers is not in the document]
Comment (thehesiod) [reply tests/test_web_urldispatcher.py:472]: I thought we decided we could keep regex for extra coverage

[the remark this answers is not in the document]
Comment (thehesiod) [reply tests/test_web_urldispatcher.py:472]: ahh add another testcase, will do

File: tests/test_web_urldispatcher.py (hunk 2 above)
Comment (thehesiod) [reply tests/test_web_urldispatcher.py]: done

**Review by webknjaz — APPROVED** [review]
